File: src/cpp/BERBitString.cpp
#include "BERBitString.h"
#include "BERCodes.h"
#include "BERException/BERExceptions.h"
#include "BERUtils.h"
#include "BasicBuffer.h"
#include <sstream>
#include <stdint.h>
#include <list>
using namespace std;

namespace BER
{
	uint32_t BERBitString::Size(BasicBuffer *value)
	{
		uint32_t result = 0;
		list<uint8_t> *tempList = BERUtils::EncodeLength(value->Size());
		result += tempList->size();
		DELETE(tempList);
		result += value->Size() / 8;
		if ((value->Size() % 8) != 0)
		{
			result++;
		}
		return result + 1;
	}

	BasicBuffer* BERBitString::Encode(BasicBuffer *value)
	{
		BasicBuffer *result = new BasicBuffer(Size(value));
		// encode the tag
		result->PushBack(BER::BitString);
		// encode the length
		list<uint8_t> *tempList = BERUtils::EncodeLength(value->Size());
		list<uint8_t>::iterator it = tempList->begin();
		while (it != tempList->end())
		{
			result->PushBack(*it++);
		}
		// encode the content
		DELETE(tempList);
		BasicBuffer *tempBasicBuffer = BERUtils::EncodeFlagList(value);
		//it = tempList->begin();
		
		//while (it != tempList->end())
		for (uint32_t i = 0; i < tempBasicBuffer->Size(); i++)
		{
			result->PushBack(tempBasicBuffer->Get(i));
		}
		DELETE(tempBasicBuffer);
		return result;
	}

	BasicBuffer* BERBitString::Decode(BasicBuffer &buffer, uint32_t &index)
	{
		BasicBuffer* result = NULL;
		if (buffer.Get(index) != BER::BitString)
		{
			throw BERTypeIncorrectTag();
		}
		index++;
		uint32_t length = BERUtils::DecodeLength(buffer, index);
		uint16_t ByteCount = 0;
		if ((length % 8) == 0)
		{
			ByteCount = length / 8;
		}
		else
		{
			ByteCount = (length / 8) + 1;
		}
		BasicBuffer *byteList = new BasicBuffer(ByteCount);
		for (uint16_t i = 0; i < ByteCount; i++)
		{
			byteList->PushBack(buffer.Get(index++));
		}
		result = BERUtils::DecodeFlagList(byteList, length);
		DELETE(byteList);
		return result;
	}
	BERBitString::~BERBitString()
	{
		DELETE(this->Value);
	}

	string BERBitString::toString()
	{

		ostringstream oss;

		oss << "{BERArray: " << (uint32_t)Value->Size() << " ";
		for (uint32_t i = 0; i < Value->Size(); i++)
		{
			bool v = Value->Get(i);
			if (v)
			{
				oss << "T";
			}
			else
			{
				oss << "F";
			}

		}
		oss << "}";
		return oss.str();

	}

}

File: src/bercs/BERBitString.cs
using System;
using System.Collections.Generic;
using System.Text;

using BER.Exceptions;

namespace BER
{
    public class BERBitString : BaseBERType<List<Boolean>>
    {
        #region Overrides of BaseBERType<List<bool>>
		//! Method that encodes the value of this BERBitString into a Byte List
		//! \return A Byte List containing the encoded raw bytes of this BERBitString
        public override List<byte> Encode(List<bool> value)
        {
            var result = new List<Byte> { (byte)BERCodes.BitString };
            result.AddRange(BERUtils.EncodeLength((uint)value.Count));
            result.AddRange(BERUtils.EncodeFlagList(value));
            return result;
        }
		//! Method that decodes a List of Bytes into this BERBitString object
		//! \return true The decoding process was succesfull
		//! \return true The decoding process was not succesfull
		//! \throws BERTypeDecodeIncorrectTag if an incompatible tag was found
        public override List<bool> Decode(List<byte> buffer)
        {
            var index = 0;
            if (buffer[index++] != (int)BERCodes.BitString)
            {
                throw new BERTypeDecodeIncorrectTag();
            }
            var temp = buffer[index++];
            UInt32 length = 0;
            int i;
            if (temp <= 0x80)
            {
                length = temp;
            }
            else
            {
                var nbrLengthBytes = temp - 0x80;
                for (i = 0; i < nbrLengthBytes; i++)
                {
                    length = (length << 8) + buffer[index++];
                }
            }
            var byteCount = Convert.ToUInt16(Math.Ceiling((Double) length/8));
            var byteList = new List<Byte>((int)byteCount);
            for (i = 0; i < byteCount; i++)
            {
                byteList.Add(buffer[index++]);
            }
            buffer.RemoveRange(0, index);
            return BERUtils.DecodeFlagList(byteList, (ushort)length);
        }
        #endregion
   		//! Method that provides a String representation of this BERBitString
		//! \return The string representation of this BERBitString as a list of boolean values represented as 0s and 1s      
        public override string AsString() { 
            StringBuilder result = new StringBuilder(Value.Count);
            foreach (bool b in Value)
                if (b) result.Append("1");
                else result.Append("0");
            return result.ToString();
        }
    }
}
